File: src/lib/bluejay/chem/integral/transform.cc
#include "bluejay/chem/integral/transform.h"
#include "bluejay/chem/integral/transform/expression.h"
#include "bluejay/core/math/einsum.h"

namespace bluejay {
namespace integral {
namespace transform {

  void IntegralTransform::transform(
    std::string expr,
    Tensor4d &t,
    const Matrix<double> &c1,
    const Matrix<double> &c2) const
  {
    auto expression = TransformExpression(expr);
    expression.Validate(t, c1, c2);
  }

  // void TransformExpression::validate(const std::vector<std::string> &symbols) const {
  //   std::set<std::string> s;
  //   s.insert(lhs.symbol);
  //   for (auto term : rhs) {
  //     if (const auto *t = boost::get<Tensor>(&term)) {
  //       s.insert(t->symbol);
  //     }
  //   }
  //   if (s != std::set<std::string>(symbols.begin(), symbols.end())) {
  //     throw std::runtime_error(
  //       "Invalid expression arguments: "
  //       "expected " + join(",", s.begin(), s.end())
  //     );
  //   }
  // }

  // template<size_t N, typename Index = size_t>
  // struct index_permutation {
  //   index_permutation(std::array<size_t,N> p, std::array<Index,N> idx)
  //     : permutation_(p), indicies_(idx)
  //   {
  //   }
  //   Index operator[](size_t i) const {
  //     return indicies_[permutation_[i]];
  //   }
  // private:
  //   std::array<size_t,N> permutation_;
  //   std::array<Index,N> indicies_;
  // };

  // void transform_2_index(
  //   std::array<size_t,4> type,
  //   IntegralEngine::TwoBodyEngine &int2e,
  //   const Molecule::Basis& basis,
  //   std::array<size_t,2> ket,
  //   Tensor4d &V)
  // {
  //   //static_assert((Type == 0213 || Type == 0123), "");
  //   for (size_t q = 0; q < basis.size(); ++q) {
  //     for (size_t p = 0; p < basis.size(); ++p) {
  //       index_permutation<4> quartet(type, { p, q, ket[0], ket[1] });
  //       auto nbf = [&] (size_t i) {
  //         return basis::nbf(basis[quartet[i]].shell());
  //       };
  //       Tensor4d v(nbf(0), nbf(1), nbf(2), nbf(3));
  //       auto shell = [&] (size_t i) {
  //         return std::tie(basis[quartet[i]].shell(), basis[quartet[i]].center());
  //       };
  //       //int2e.compute(shell(0), shell(1), shell(2), shell(3), v);
  //     }
  //   }
  // }


  void IntegralTransform::compute_pq(size_t k, size_t l) {
    // const auto &basis = this->basis_;
    // // const auto &r = basis[k];
    // // const auto &s = basis[l];
    // for (const auto &p : basis) {
    //   for (const auto &q : basis) {
    //     //Tensor4d eri;
    //     //integral_engine_->compute(p, q, r, s, eri);
    //     //eri_(r,s,p,q) = eri;
    //   }
    // }
  }

  Tensor4d IntegralTransform::transform_pq(const Matrix<double> &Ci, const Matrix<double> &Cj) const {
    // const auto &V = this->eri_;
    // Tensor4d t1, t2;
    // einsum::einsum(Cj, 1, V, 3, t1); // u(j,r,s,p) = C(j,q)*v(r,s,p,q)
    // einsum::einsum(Ci, 1, V, 3, t2); // t(i,j,r,s) = C(i,p)*u(j,r,s,p)
    //return t2;
    return Tensor4d{};
  }

  Tensor4d IntegralTransform::transform_pq(const Tensor4d &Tij) const {
    //const auto &V = this->eri_;
    //auto t2 = einsum("ijpq,rspq", Tij, V); // t(i,j,r,s) = T(i,j,p,q)*v(r,s,p,q)
    return Tensor4d{};
  }

}
}
}

File: src/lib/python/bluejay/chem/integral/transform.py
import bluejay.chem

class TransformExpression:

  @staticmethod
  def parse(expr):
    print ('TransformExpression.parse(%r)' % expr)
    import pyparsing as pp
    pp_version = ([ int(s) for s in pp.__version__.split(".") ])
    assert [2,2,0] <= pp_version < [2, 3, 0]
    #pp.ParserElement.setDefaultWhitespaceChars(' \t')
    #symbol = pp.Word(pp.alphas)('symbol')
    symbol = pp.pyparsing_common.identifier
    index = pp.Word(pp.alphas, exact=1)
    index_list = pp.delimitedList(index, ',')
    braket = '(' + index_list('bra') + '|' + index_list('ket') + ')'
    tensor = (symbol('symbol') + '(' + index_list('index_list') + ')')
    transform = tensor('transforms*')
    lhs = tensor("lhs")
    parser = lhs + '=' \
             + (pp.ZeroOrMore(transform + '*')) \
             + braket \
             + (pp.ZeroOrMore('*' + transform))
    result = parser.parseString(expr)
    #print (result.dump())
    bra = list(result.bra)
    ket = list(result.ket)
    from collections import namedtuple
    symbol_index_tuple = namedtuple('symbol_index_tuple', 'symbol index_list')
    args = [
      symbol_index_tuple(a.symbol, list(a.index_list))
      for a in [ result.lhs, *result.transforms ]
    ]
    #print (result.lhs.dump())
    return (bra,ket,args)

  @staticmethod
  def assert_unique(seq,msg):
    dup = set([ s for s in seq if seq.count(s) > 1 ])
    if dup: raise Exception(msg.format(dup))


  def __init__(self, expr, **kwargs):

    self._expr = expr
    self._transforms = []

    (bra,ket,args) = TransformExpression.parse(self._expr)
    # print ('bra,ket = %s %s' % (bra, ket))
    # print ('args = %s' % (args))

    symbols = set([ symbol for (symbol,index) in args ])
    if symbols != kwargs.keys():
      raise Exception(
        'Expression symbols %s do not match kwargs %s'
        % (symbols, set(kwargs.keys()))
      )

    braket = bra+ket
    self.assert_unique(braket, 'Duplicate braket indices: {0}')

    index_mapping = { label : index for (index,label) in enumerate(sorted(braket)) }
    #print ('index mapping:', index_mapping)
    self._braket = [ index_mapping[idx] for idx in braket ]

    free_indices = []
    transformed_indices = []

    for (symbol,index_list) in args[1:]:
      free = [ idx for idx in index_list if idx not in braket ]
      dummy = [ idx for idx in index_list if idx in braket ]
      if not (free and dummy):
        raise Exception(
          'Invalid transform %s(%s): free=%s, dummy=%s' %
          (symbol, ','.join(index_list), free, dummy)
        )
      #print ('transform: ', symbol, free, dummy)
      free_indices += free
      transformed_indices += dummy
      transform = (kwargs[symbol],) + tuple((index_list.index(idx), index_mapping[idx]) for idx in dummy)
      #print ("transform:", transform)
      self._transforms.append(transform)

    self.assert_unique((free_indices+transformed_indices), 'Duplicate rhs indices: {0}')
    self.assert_unique(args[0].index_list, 'Duplicate rhs indices: {0}')

    lhs = args[0].index_list
    rhs = free_indices + [ idx for idx in braket if idx not in transformed_indices ]
    if sorted(lhs) != sorted(rhs):
      raise Exception(
        'Left-hand side indices %s are not equivalent to right-hand side %s',
        lhs, rhs
      );

    self._lhs = kwargs[args[0].symbol]
    self._order = [ rhs.index(idx) for idx in lhs]



class IntegralTransform:

  def __init__(self, basis, integral_engine=None):
    integral_engine = integral_engine or bluejay.chem.integral_engine()
    self._basis = basis
    self._integral_engine_ = integral_engine
    # self._transform = bluejay_chem.IntegralTransform(basis, integral_engine)
    # self._transforms = []

  def compute(self, expr, **kwargs):
    import bluejay_chem.integral, bluejay_core
    expr = TransformExpression(expr, **kwargs)
    # print (
    #   "bluejay_chem.integral.transform: braket=%s, lhs=%s, order=%s, %s"
    #   % (expr._braket, expr._lhs, expr._order, expr._transforms)
    # )

    t = bluejay_chem.integral.IndexTransform(
      #bluejay_core.Array(expr._lhs),
      expr._lhs,
      expr._order, *expr._transforms
    )

    bluejay_chem.integral.transform(
      self._basis,
      self._integral_engine_.coulomb(),
      expr._braket,
      t
    )

    #   bluejay_chem.integral.IndexTransform(
    #     #bluejay_core.Array(expr._lhs),
    #     expr._lhs,
    #     expr._order, *expr._transforms)
    # )
